ee/hogai/root/nodes.py

```diff
@@ -80,55 +99,112 @@ def run(self, state: AssistantState, config: RunnableConfig) -> PartialAssistant
-            message.tool_call_id: message for message in state.messages if isinstance(message, AssistantToolCallMessage)
+            message.tool_call_id: message
+            for message in conversation_window
+            if isinstance(message, AssistantToolCallMessage)
         }
 
         history: list[BaseMessage] = []
-        for message in state.messages:
+        for message in conversation_window:
             if isinstance(message, HumanMessage):
-                history.append(LangchainHumanMessage(content=message.content))
+                history.append(LangchainHumanMessage(content=message.content, id=message.id))
             elif isinstance(message, AssistantMessage):
                 # Filter out tool calls without a tool response, so the completion doesn't fail.
                 tool_calls = [
                     tool for tool in message.model_dump()["tool_calls"] or [] if tool["id"] in tool_result_messages
                 ]
 
-                history.append(LangchainAIMessage(content=message.content, tool_calls=tool_calls))
+                history.append(LangchainAIMessage(content=message.content, tool_calls=tool_calls, id=message.id))
 
                 # Append associated tool call messages.
                 for tool_call in tool_calls:
+                    tool_call_id = tool_call["id"]
+                    result_message = tool_result_messages[tool_call_id]
                     history.append(
                         LangchainToolMessage(
-                            content=tool_result_messages[tool_call["id"]].content, tool_call_id=tool_call["id"]
+                            content=result_message.content, tool_call_id=tool_call_id, id=result_message.id
                         )
                     )
 
+        # Find a new window id and trim the history to it.
+        new_window_id = self._find_new_window_id(state, history)
+        if new_window_id is not None:
+            history = self._get_conversation_window(history, new_window_id)
+
         if self._is_hard_limit_reached(state):
             history.append(LangchainHumanMessage(content=ROOT_HARD_LIMIT_REACHED_PROMPT))
 
-        return history
+        return history, new_window_id
 
     def _is_hard_limit_reached(self, state: AssistantState) -> bool:
         return state.root_tool_calls_count is not None and state.root_tool_calls_count >= self.MAX_TOOL_CALLS
 
+    def _find_new_window_id(self, state: AssistantState, window: list[BaseMessage]) -> str | None:
+        """
+        If we simply trim the conversation on N tokens, the cache will be invalidated for every new message after that
+        limit leading to increased latency. Instead, when we hit the limit, we trim the conversation to N/2 tokens, so
+        the cache invalidates only for the next generation.
+        """
+        if settings.TEST:
+            model = ChatOpenAI(model="gpt-4o", api_key="token_counter")
+        else:
+            model = self._get_model(state)
```

Comment:
This is a little smelly – why not _mock_ `_get_model()` in tests instead, so that prod code is not aware of this?